src/ClientModel/SamplingDataPostProcessor.cpp

```diff
@@ -171,47 +188,51 @@ void SamplingDataPostProcessor::SortByThreadUsage() {
-
-        for (uint64_t address : callstack.frames()) {
-          if (exact_address_to_function_address_.find(address) ==
-              exact_address_to_function_address_.end()) {
-            MapAddressToFunctionAddress(address, capture_data);
-          }
-          uint64_t function_address = exact_address_to_function_address_.at(address);
+  callstack_data.ForEachUniqueCallstack([this, &capture_data](uint64_t callstack_id,
+                                                              const CallstackInfo& callstack) {
+    // A "resolved callstack" is a callstack where every address is replaced by the start address of
+    // the function (if known).
+    std::vector<uint64_t> resolved_callstack_frames;
+
+    for (uint64_t address : callstack.frames()) {
+      if (exact_address_to_function_address_.find(address) ==
+          exact_address_to_function_address_.end()) {
+        MapAddressToFunctionAddress(address, capture_data);
+      }
+      uint64_t function_address = exact_address_to_function_address_.at(address);
+      resolved_callstack_frames.push_back(function_address);
+    }
 
-          resolved_callstack_frames.push_back(function_address);
-          function_address_to_sampled_callstack_ids_[function_address].insert(callstack_id);
-        }
+    // For non-kComplete callstacks, only use the innermost frame for statistics.
+    if (callstack.type() == CallstackInfo::kComplete) {
+      for (uint64_t function_address : resolved_callstack_frames) {
+        function_address_to_sampled_callstack_ids_[function_address].insert(callstack_id);
+      }
+    } else {
+      function_address_to_sampled_callstack_ids_[resolved_callstack_frames[0]].insert(callstack_id);
+    }
 
-        // Check if we already have this callstack
-        uint64_t resolved_callstack_id;
-        auto it = resolved_callstack_to_id_.find(resolved_callstack_frames);
-        if (it == resolved_callstack_to_id_.end()) {
-          resolved_callstack_id = callstack_id;
-          CHECK(!id_to_resolved_callstack_.contains(resolved_callstack_id));
-          orbit_client_protos::CallstackInfo resolved_callstack;
-          *resolved_callstack.mutable_frames() = {resolved_callstack_frames.begin(),
-                                                  resolved_callstack_frames.end()};
-          resolved_callstack.set_type(callstack.type());
-          id_to_resolved_callstack_.insert_or_assign(resolved_callstack_id, resolved_callstack);
-          resolved_callstack_to_id_.insert_or_assign(std::move(resolved_callstack_frames),
-                                                     resolved_callstack_id);
-        } else {
-          resolved_callstack_id = it->second;
-        }
+    // Check if we already have this resolved callstack, and if not, create one.
+    uint64_t resolved_callstack_id;
+    auto it =
+        resolved_callstack_to_id_.find(std::make_pair(resolved_callstack_frames, callstack.type()));
```

Comment:
This is kind of expensive because that pair is only created for the lookup. To create that pair the vector has to be copied.
I'm not sure whether performance matters here, but this could be solved with heterogeneous lookup.

If you want to tackle this I can provide an example how this could be solved.

Reply:
You're totally right, this was a bit naive. I added a separate commit tackling this (it's quite a lot of boilerplate but oh well). Please have a look and let me know if what I did makes sense.

Reply:
Looks good to me. Sorry for inflicting so much boiler plate. :grimacing:

Reply:
Thanks for catching it!